### PR TITLE
Add a new exporter type: DirectFLP

### PR DESCRIPTION
**Pull request by jotak**
## Description

DirectFLP starts the whole flowlogs-pipeline process directly from the
Agent, eliminating the need for a network connection between them.

This can be useful when both Agent and FLP run on the same host, saving
resources. This is mutually exclusive with other exporters such as with
Kafka.

## Dependencies

https://github.com/netobserv/flowlogs-pipeline/pull/538

(hold until this is merged => replace directive to remove)

## Checklist

If you are not familiar with our processes or don't know what to answer in the list below, let us know in a comment: the maintainers will take care of that.

* [ ] Will this change affect NetObserv / Network Observability operator? If not, you can ignore the rest of this checklist.
* [ ] Is this PR backed with a JIRA ticket? If so, make sure it is written as a title prefix _(in general, PRs affecting the NetObserv/Network Observability product should be backed with a JIRA ticket - especially if they bring user facing changes)._
* [ ] Does this PR require product documentation?
  * [ ] If so, make sure the JIRA epic is labelled with "documentation" and provides a description relevant for doc writers, such as use cases or scenarios. Any required step to activate or configure the feature should be documented there, such as new CRD knobs.
* [ ] Does this PR require a product release notes entry?
  * [ ] If so, fill in "Release Note Text" in the JIRA.
* [ ] Is there anything else the QE team should know before testing? E.g: configuration changes, environment setup, etc.
  * [ ] If so, make sure it is described in the JIRA ticket.
* QE requirements (check 1 from the list):
  * [ ] Standard QE validation, with pre-merge tests unless stated otherwise.
  * [ ] Regression tests only (e.g. refactoring with no user-facing change).
  * [x] No QE (e.g. trivial change with high reviewer's confidence, or per agreement with the QE team).

### CODE REVIEW

File: pkg/agent/config.go [hunk 1]
```diff
@@ -140,6 +140,9 @@ type Config struct {
 	KafkaSASLClientSecretPath string `env:"KAFKA_SASL_CLIENT_SECRET_PATH"`
 	// ProfilePort sets the listening port for Go's Pprof tool. If it is not set, profile is disabled
 	ProfilePort int `env:"PROFILE_PORT"`
+	// Flowlogs-pipeline configuration, used when export is "direct-flp". Cf https://github.com/netobserv/flowlogs-pipeline
+	// The "ingest" stage must be omitted from this configuration, since it is handled internally by the agent.
+	FLPConfig string `env:"FLP_CONFIG"`
```
Comment (jpinsonneau) [inline pkg/agent/config.go:145]: As soon as `FLP_CONFIG` is set, `EXPORT` should be ignored.
The export capabilities from pod point of view are inside FLP config right ?

File: pkg/agent/agent.go [hunk 2]
```diff
@@ -251,6 +251,8 @@ func buildFlowExporter(cfg *Config) (node.TerminalFunc[[]*flow.Record], error) {
 		return buildIPFIXExporter(cfg, "udp")
 	case "ipfix+tcp":
 		return buildIPFIXExporter(cfg, "tcp")
+	case "direct-flp":
+		return buildDirectFLPExporter(cfg)
```
Comment (jpinsonneau) [inline pkg/agent/agent.go:255]: I would move that outside of cfg.Export as:

```
if len(cfg.FLPConfig) {
  return buildDirectFLPExporter(cfg)
}
switch cfg.Export {
  ...
}
```

**Review by jpinsonneau — COMMENTED** [review]
Approach sounds good ! Just a small change on config to make it clear for the user
We should also add an example in `deployments` folder

File: pkg/agent/config.go (hunk 1 above)
Comment (jotak) [reply pkg/agent/config.go:145]: I guess it depends from where you sit.. In my mind, this new mode is really a new exporter, sitting from the Agent codebase PoV: the agent generates flows and then exports them somewhere, here to FLP; the fact that FLP is in the same process is a detail, it's still an exporter from this codebase PoV.

I guess you have more the "external user point of view" where, once FLP config is set, "export" would mean what happens "after FLP processing", which indeed doesn't really make sense since FLP does the export (to loki, prom, etc.)

idk, I think both approaches can be defended ... @msherif1234 , some preference?

Comment (msherif1234) [reply pkg/agent/config.go:145]: thinking of it as new exporter mode resonate well with me as well

File: pkg/exporter/direct_flp.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	*ingest.InProcess
+}
+
+func StartDirectFLP(jsonConfig string) (*DirectFLP, error) {
+	var cfg flpconfig.ConfigFileStruct
+	// Note that, despite jsonConfig being json, we use yaml unmarshaler because the json one
+	// is screwed up for HTTPClientConfig in github.com/prometheus/common/config (used for Loki)
+	// This is ok as YAML is a superset of JSON.
+	// E.g. try unmarshaling `{"clientConfig":{"authorization":{}}}` as a api.WriteLoki
+	// See also https://github.com/prometheus/prometheus/issues/11816
```
Comment (msherif1234) [inline pkg/exporter/direct_flp.go:24]: interesting thanks for referring to the issue in promo its almost 1yr so probably this work around will stay for sometime

File: pkg/exporter/direct_flp.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+	return &DirectFLP{InProcess: flp}, nil
+}
+
+// ExportFlows accepts slices of *flow.Record by its input channel, converts them
+// to *pbflow.Records instances, and submits them to the collector.
+func (d *DirectFLP) ExportFlows(input <-chan []*flow.Record) {
+	for inputRecords := range input {
+		pbRecords := flowsToPBNoChunk(inputRecords)
+		d.Write(pbRecords)
```
Comment (msherif1234) [inline pkg/exporter/direct_flp.go:42]: This step can be removed if we let the new FLP pipe process ebpf record directly not protobuf records

File: pkg/exporter/direct_flp.go (hunk 3 above)
Comment (jotak) [reply pkg/exporter/direct_flp.go:24]: Yeah there's even a PR from @OlivierCazade that is older than that :-) https://github.com/prometheus/common/pull/394

File: pkg/exporter/direct_flp.go (hunk 4 above)
Comment (jotak) [reply pkg/exporter/direct_flp.go]: done via [bcccbb7](https://github.com/netobserv/netobserv-ebpf-agent/pull/223/commits/bcccbb7dfca1b8ec9f1e8940911dce6c59376660)

File: pkg/decode/decode_protobuf.go [hunk 5]
```diff
@@ -67,26 +67,30 @@ func PBFlowToMap(flow *pbflow.Record) config.GenericMap {
...
-		if proto == syscall.IPPROTO_TCP || proto == syscall.IPPROTO_UDP || proto == syscall.IPPROTO_SCTP {
-			if proto == syscall.IPPROTO_TCP {
-				out["SrcPort"] = flow.Transport.GetSrcPort()
-				out["DstPort"] = flow.Transport.GetDstPort()
-				out["Flags"] = flow.Flags
-			} else {
-				out["SrcPort"] = flow.Transport.GetSrcPort()
-				out["DstPort"] = flow.Transport.GetDstPort()
-			}
+	if ethType != ethernet.EtherTypeIPv4 && ethType != ethernet.EtherTypeIPv6 {
```
Comment (jotak) [inline pkg/decode/decode_protobuf.go:71]: @msherif1234 I'm doing this early return here when it's not ipv4/ipv6, could you confirm this is correct?
Previously, in that case (e.g. with ARP), the code would still check for drops/DNS/RTT fields and now this isn't the case anymore, but I believe our drops/DNS/RTT metrics wouldn't make sense for non-IP packets, right?

If this doesn't sound good to you, I can rollback, but I'd need to adapt also the ConvertToFLP function because I'm enforcing in my tests that the same flow input produces the exact same output regardless if it's using ConvertToFLP or converting via protobuf.

File: pkg/exporter/convert_flp_test.go [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+	"google.golang.org/protobuf/proto"
+)
+
+func TestConvertFLP(t *testing.T) {
+	decoder := decode.Protobuf{}
+
+	someTime := time.Now()
+	var someDuration time.Duration = 10000000 // 10ms
+
+	tests := []struct {
```
Comment (jotak) [inline pkg/exporter/convert_flp_test.go:24]: @msherif1234 FYI the tests here are the same that were previously in `decode_protobuf_test` except that they're now used to test both pb conversion and direct FLP conversion, making sure they both produce the same result

File: pkg/decode/decode_protobuf.go (hunk 5 above)
Comment (msherif1234) [reply pkg/decode/decode_protobuf.go:71]: the only thing we could be missing here is l2 drops, also pls test with DNS TCP to be sure `DNSError` is only none zero for TCP handshake I can help you with this test if u like

File: pkg/decode/decode_protobuf_test.go [hunk 7]
```diff
@@ -7,375 +7,11 @@ import (
...
-		rawPB, err := proto.Marshal(tt.flow)
-		require.NoError(t, err)
-		out, err := decoder.Decode(rawPB)
-		require.NoError(t, err)
-		assert.NotZero(t, out["TimeReceived"])
-		delete(out, "TimeReceived")
-		assert.Equalf(t, *tt.expected, out, tt.name)
-	}
-}
+// NOTE: more tests in convert_flp_test.go
```
Comment (msherif1234) [inline pkg/decode/decode_protobuf_test.go:14]: did intentionally deleted conversion from gRPC to generic map test cases here ?

File: pkg/exporter/convert_flp_test.go (hunk 6 above)
Comment (msherif1234) [reply pkg/exporter/convert_flp_test.go:24]: this test coverting from ebpf record to flp configmap but it seems u deleted from gRPC to configmap in case this mode isn't set ?

File: pkg/decode/decode_protobuf_test.go (hunk 7 above)
Comment (jotak) [reply pkg/decode/decode_protobuf_test.go:14]: yes, like I wrote this test was moved (and sightly changed) to convert_flp_test, but I admit the name is misleading, I'll rename that file to clarify it's testing BOTH grpc and direct-flp conversions

File: pkg/exporter/convert_flp_test.go (hunk 6 above)
Comment (jotak) [reply pkg/exporter/convert_flp_test.go:24]: cf my next comment below

File: pkg/exporter/convert_flp_test.go [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+	for _, tt := range tests {
+		// Test direct conversion
+		outDirect := ConvertToFLP(tt.flow)
+		assert.NotZero(t, outDirect["TimeReceived"])
+		delete(outDirect, "TimeReceived")
+		err := normalizeMap(outDirect)
+		require.NoError(t, err)
+		assert.Equalf(t, *tt.expected, outDirect, tt.name)
+
+		// Test pb conversion
```
Comment (jotak) [inline pkg/exporter/convert_flp_test.go:348]: see here @msherif1234 : I am testing both FLP conversion AND protobuf conversion, with the bonus thatit makes sure the two are producing the same result (they're both compared against the same expected map)

**Comment by jotak** [comment]
thanks @msherif1234 !
/approve
